storage/innobase/clone/clone0copy.cc

```diff
@@ -1389,6 +1427,19 @@ int Clone_Handle::copy(THD *thd, uint task_id, Ha_clone_cbk *callback) {
         break;
       }
 
+      // TODO(laurynas): this should be before the send_state_metadata call
+      // above, and the current placement would result in precopy (but not
+      // engine synchronization) skipped if restarted at an unfortunate time.
+      // But placing it there desyncs the clone protocol, which is not debugged
+      // yet. And skipping precopy is not a correctness issue, but a performance
+      // one.
+      if (in_precopy_state) {
+        err = callback->precopy(thd, task_id);
```

Comment:
It seems like the work performed here is similar to m_clone_task_manager.reserve_next_chunk() / process_chunk() which happens before the send_state_metadata(). I wonder if this is causing your restart issues? I'm not sure I understand what you mean by "desyncs the clone protocol"?

Reply:
> I'm not sure I understand what you mean by "desyncs the clone protocol"?

Both client and donor expect the other side to send something